docs/USER_RECORD.md

```diff
@@ -314,7 +314,20 @@ user. When logging in
 [`pam_systemd`](https://www.freedesktop.org/software/systemd/man/pam_systemd.html)
 will automatically initialize the `$LANG` environment variable from this
 string. The string hence should be in a format compatible with this environment
-variable, for example: `de_DE.UTF8`.
+variable, for example: `de_DE.UTF-8`. You should use `languages` instead.
+
+`languages` → An array of strings indicating the preferred languages/locales for
+the user, listed in order of descending priority. This allows multi-lingual users
+to specify all the languages that they know, so software lacking translations in
+the user's primary language can fall back to another language that the user knows
+rather than the default English. When logging in
+[`pam_systemd`](https://www.freedesktop.org/software/systemd/man/pam_systemd.html)
+will automatically initialize the `$LANGUAGE` variable from this list. `pam_systemd`
+will also initialize the `$LANG` environment variable with the first entry in this
+array (i.e. the user's primary language), ignoring `preferredLanauge`. All the entires
+in the array must be in a format compatible with the `$LANG` environment variable,
+for example: `de_DE.UTF-8`. If `preferredLanguage` is set, it must be set to the first
+entry of `languages`.
```

Comment:
The behavior of having both `LANGUAGE` and `LANG` set, to conflicting values, isn't documented and is thus undefined.

So either LANGUAGE and LANG need to be set to matching values, or only LANG can be set. Setting only LANGUAGE makes no sense because it does nothing unless LANG is set also.

The fields reflect this. We either set `languages` (which sets both LANGUAGE and LANG) and require that `preferredLanguage` (i.e. LANG) is either default (free to come from `languages`) or already matching `languages`. Or if we're looking at a user record created before `languages` existed, we will just have `preferredLanguage` (i.e. only LANG)

> I have the suspicion 99% of users just want one language

Sure. But we can represent this w/ an array of one element :)

Reply:
Honestly, it would be nice to be able to make the occasional breaking changes to user records. Some versioning system?

We should also probably make a couple big breaking changes now, before much software is made compatible with user records. Stuff like cleaning up duplicate/deprecated fields (like `preferredLanguage` is, in my books, kinda useless). Reworking how we store the various authentication methods with objects, so we can add extra fields (where it's stored in the LUKS header, an encrypted version of the recovery key so that we can show it in gnome-settings, etc)

Reply:
So far we have been very good with avoiding breaking changes, and this one doesn't look like it's strong enough to start with this.

Given, that the first item in the languages array is special anyway, I'd really just say that preferredLanguage *is* considered to be just the first item of the array, if it is set.

so I'd propose a logic like this on login: generate a single array of `preferredLanguage` as first item, and `language` as suffix. Then remove duplicates. Then set $LANG to the first item of the resulting array, and $LANGUAGE to the whole array.

This would have the effect we have perfect compat with the status quo ante, we set both env vars to something useful, and we removed most pitfalls I think, because both fields are always taken into consideration.

The merging/uniquifying of the two fields I'd hide in a user_record_get_languages() accessor or so.

Reply:
> and this one doesn't look like it's strong enough to start with this

Definitely this isn't important enough. There's just some cruft that would be nice to get rid of at some point before everything starts using user records

> so I'd propose a logic like this on login

Sounds great

Reply:
Hmm I think it makes sense to rename the `languages` variable to `fallbackLanguages` then

Reply:
well, "fallback" to me sounds as if we'd just pick one of them. But this is not how this works afaics. they are *all* used, i.e. for each translatabnle string a search is done among the languages, no?

Maybe "additionalLanguages"?